src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs

```diff
@@ -65,38 +78,79 @@ private static void Scan(object state)
 
         private void Scan()
         {
-            // Scan the registered connections looking for ones that have timed out
-            foreach (var c in _connections)
+            if (_running)
             {
-                if (!c.Value.Active && (DateTimeOffset.UtcNow - c.Value.LastSeenUtc).TotalSeconds > 5)
+                return;
+            }
+
+            try
+            {
+                _running = true;
```

Comment:
If this can actually happen simultaneously, you need `cmpexchg`. There's a race here otherwise. But I'm not sure how it can happen simultaneously, is it just in situations where the `Scan` takes more than one Timer interval? If so, suspend the timer when you start or something.